pkg/cover/canonicalizer.go

```diff
@@ -82,11 +88,13 @@ func (can *Canonicalizer) NewInstance(modules []host.KernelModule) *Canonicalize
 		canonicalToInstMap[canonicalAddr] = &canonicalizerModule{
 			offset:  int(instAddr) - int(canonicalAddr),
 			endAddr: uint32(module.Size) + canonicalAddr,
+			discard: discard,
```

Comment:
Could the following implementation be simpler?

1) If we decide to discard the module, just don't add it to maps.
2) In `ConvertPC`, only keep PCs that match modules known to the `CanonicalizerInstance`. The kernel itself is also stored as one, right?

Reply:
The kernel itself is currently not stored as a module, so if the module is discarded any addresses falling into this category will just be converted like a typical kernel value. I'll look into storing it as a module\

Reply:
The kernel itself is stored as a module when creating the coverage report generator: https://github.com/google/syzkaller/blob/2a0d0f29e37863d5cdb8e95c9fa2a83251ba3972/pkg/cover/backend/modules.go#L21-L24

However, the kernel is not stored as a module in syz-manager's module list, as the module list used by syz-manager is generated directly from proc/modules:
https://github.com/google/syzkaller/blob/2a0d0f29e37863d5cdb8e95c9fa2a83251ba3972/pkg/host/machine_info_linux.go#L128-L131

It would be doable to store the kernel as a module by running the report generator before canonicalization, then going through all symbols and finding the last one with the main kernel as a module
https://github.com/google/syzkaller/blob/2a0d0f29e37863d5cdb8e95c9fa2a83251ba3972/pkg/cover/backend/backend.go#L33-L38

This may be more accurate as we can drop all coverage that doesn't belong to a known module. However this would require generating a coverage report early, and we could lose valid coverage if it falls outside of a known symbol (though I'm not sure if that is possible/a concern)

Reply:
Thank you for the analysis!

Indeed, it will only complicate things. Let's follow the current `discard bool` approach then.